### DISCUSSION TITLE
Importing AST_NODE_TYPES from @typescript-eslint/typescript-estree is very slow

### DISCUSSION THREAD

**Discussion by ericmorand** [typescript-eslint/typescript-eslint#9613] [Community Feedback]
Hey maintainers,

A few days ago, while I was working on [One Double Zero](https://www.npmjs.com/package/one-double-zero), I noticed that running the `version` command of the application was very slow - somewhere around half a second to get the version of the application output to the console.

After investigating, I discovered that importing `AST_NODE_TYPES` from @typescript-eslint/typescript-estree was very slow - something around 350ms just to execute `import {AST_NODE_TYPES} from "@typescript-eslint/typescript-estree";` at runtime. After digging into the source code of the package, I discovered that `@typescript-eslint/typescript-estree`, at some point, import the whole typescript package before it is needed - i.e. typically here, when importing `AST_NODE_TYPES`, there is no need for typescript to be imported.

I had to move the import of the package [into the factory of my own custom parse function](https://gitlab.com/nightlycommit/one-double-zero/-/blob/main/src/main/domain/ast.ts?ref_type=heads#L46), and had to start importing `AST_NODE_TYPES` from `@typescript-eslint/types` instead, which solved the issue...until today where I imported again and without noticing `AST_NODE_TYPES` from `@typescript-eslint/typescript-estree` - thank you dear IDE.

I'm about to release a version that fix the issue one more time, but I'm sure that it will happen again, and I'm also sure that a lot of people are actually doing the same mistake without even noticing.

I'd like to request that either:

* `@typescript-eslint/typescript-estree` stops exporting AST_NODE_TYPES - to make sure that IDEs only import it from `@typescript-eslint/types`
* or that AST_NODE_TYPES is removed entirely and node types are replaced by strings (`type: 'WhileStatement'`), removing the dependency on an enum entirely while preserving type checking and autocompletion
* or that the typescript package is imported only when needed instead of as soon as `@typescript-eslint/typescript-estree` is imported

**Comment by bradzacher** [comment]
> before it is needed

What do you mean "before it is needed"?
If you're accessing `AST_NODE_TYPES` it's because you've got an AST you're analysing.
If you've got an AST you're analysing then you should have already loaded TypeScript -- because our parser is built directly on top of TypeScript.

> or that AST_NODE_TYPES is removed entirely and node types are replaced by strings (type: 'WhileStatement')

Not going to happen. You're free to use strings if you want. The enum is there for good reason because it prevents whole classes of issues and makes things more discoverable.

> @typescript-eslint/typescript-estree stops exporting AST_NODE_TYPES - to make sure that IDEs only import it from @typescript-eslint/types

You're free to restrict the module within your codebase using lint rules like
https://eslint.org/docs/latest/rules/no-restricted-imports#importnames

Why would we change our API surface and prevent people from accessing `AST_NODE_TYPES` via the parser they are using?

**Comment by ericmorand** [comment]
> You're free to use strings if you want.

Now that you mention it, that's the best way to solve the issue: instead of using AST_NODE_TYPES, I'll use the string themselves. Theorically speaking, it is faster (although negligibly) since we don't have to access an object member.

Thanks for the tip.
